Remove deprecated functions 14

bangerth opened the pull request:
This removes the naked compress() function without VectorOperation argument from the Trilinos classes.
Fixes #403

Rombur commented on source/lac/trilinos_block_sparse_matrix.cc:
```diff
@@ -258,7 +258,7 @@ namespace TrilinosWrappers
   void
   BlockSparseMatrix::collect_sizes ()
   {
-    compress();
```
Why don't you need this compress anymore ?

bangerth replied:
I think that in this case I really don't need it because it doesn't actually access individual elements of the matrix. It only accesses the sizes of the individual blocks which are never affected by `compress` on the blocks.

Rombur commented on source/lac/trilinos_sparse_matrix.cc:
```diff
@@ -126,9 +126,6 @@ namespace TrilinosWrappers
 
-      // otherwise first flush Trilinos caches
-      matrix->compress ();
-
```
Why don't you need this compress anymore ?

bangerth replied:
OK, this is actually an interesting case. The code we previously had was actively wrong in that if you looped over all of the locally owned rows, you'd call `compress()` for each of these rows. But that means calling `compress()` a different number of times on each processor. This may have worked by accident if individual processors (wrongly) decide to elide the communication if nothing had been changed locally, but it would fail as soon as one processor has no rows at all.

Furthermore, it is extremely inefficient. If you loop over the locally owned elements of a parallel matrix (or, just all elements of a non-parallel matrix), then this calls `compress()` many times even though it is likely not going to make any kind of difference.

In other words, I think the call here was inserted because it seemed like the safe thing to do, but not because it was right. That said, I'm going to add a notice to the `begin()` functions where we create these iterators stating that the matrix needs to be "clean" before using iterators on it.

bangerth replied:
I'm going to add another test `mpi/trilinos_sparse_matrix_03` that verifies the semantics one would expect in this context. It actually points to another problem, namely that one can call `begin(row)` on a locally owned row, but not `end(row)` if `row` is the last row the current processor owns -- awkward...

bangerth replied:
That actually took a while to fix, but there are now `mpi/trilinos_sparse_matrix_0[45]` for this.

Rombur commented on source/lac/trilinos_sparse_matrix.cc:
```diff
@@ -1102,7 +1099,6 @@ namespace TrilinosWrappers
   SparseMatrix::clear_rows (const std::vector<size_type> &rows,
                             const TrilinosScalar          new_diag_value)
   {
-    compress();
```
This will change how this function works. A code that was valid before may now throw an exception.

bangerth replied:
This, too, is interesting. `clear_rows()` calls `clear_row()` for each of the rows, but `clear_row` ignores all rows not locally stored. `clear_rows()` called `compress()` at the beginning (it can't know whether any add or set operations are pending, so it doesn't know what the correct operation is) and then calls `compress(insert)` again at the end. The latter clearly makes no sense because every processor only operates on its local elements and there are no ghost rows for matrices, so there is nothing to communicate. I feel that the former is wrong because we would need to know what operation is pending but can't right here. I'll document that this is the user's responsibility.

Rombur commented on source/numerics/matrix_tools.cc:
```diff
@@ -2463,8 +2463,6 @@ namespace MatrixTools
     const unsigned int n_blocks = matrix.n_block_rows();
 
-    matrix.compress();
-
```
Same as above.

bangerth replied:
As before, I don't think this can work reliably. I've documented the issue.

Rombur commented on source/numerics/matrix_tools.cc:
```diff
@@ -2559,12 +2557,6 @@ namespace MatrixTools
-            // work if we compress the matrix first,
-            // done here
-            matrix.compress ();
-
```
With this change and the change in clear_rows, code that was valid before can now throw an exception.

bangerth replied:
Same here. I think this was particularly insidious: it happens in a block guarded by `if (boundary_values.size()>0)` which means that interior processors did not call the function and it would probably have resulted in a deadlock in that case.

Rombur commented:
I think that you are changing the behaviour of some functions. I think that's OK but then, it should be put as incompatible changes in change.txt.

bangerth commented:
These are totally reasonable questions, and maybe I should have discussed this more in the pull request.

Most of the calls to `compress()` that I removed are at the beginning of functions. In other words, the old functions assumed nothing about the state of the matrix whereas after the change, the function assumes that there are no pending writes or additions to the matrix. The reason I removed these calls are:
- If you call `compress` with a `VectorOperation` argument, you need to know whether the last operation was a write or an addition. We cannot know this at the beginning of a function since these operations would have happened in other functions upstream somewhere out of the control of the current function. In other words, we cannot reasonably call `compress()` in these places unless the matrix keeps track what operation was last performed.
- It's pretty inefficient. Some of these places are just short read accesses that might happen in a loop and we probably don't want to call `compress()` in each iteration of the loop.
- It might lead to deadlocks. `compress()` is a collective operation, so every processor needs to execute the same number of calls (i.e., also the same number of iterations through the loop). This is, for example, why the `clear_row()` function did not have a `compress()` call in it, but `clear_rows()` does -- the former is called an unpredictable number of times on each processor, whereas the latter is called exactly once.

I should probably document the functions where I have removed the call to `compress()` to state that the matrix needs to have `compress()` called beforehand as appropriate, but I'd also appreciate a more general discussion whether the approach taken here is reasonable or not.

Rombur commented:
I don't have a strong opinion on this. I think that the ```last_operation``` variable keeps track of the last operation performed and we could check the ```compressed``` variable before we call compress. If you prefer to let the user call ```compress()```, it's fine with me. We should just warn about a possible incompatibility.

kronbichler commented:
I agree with @bangerth that this is a good way forward. last_operation cannot synchronize between processors and at least I felt that the design that stores these things locally was brittle when I wrote the code in 2008. (It was taken from the PETSc wrappers). So requiring the user to explicitly call compress with the appropriate action is definitely better.

tjhei commented:
The policy I am favoring and now enforcing inside PETSc is the following:
1. The user is expected to call compress() with the correct argument after manually operating on individual elements.
2. The user never needs to call compress() in a different situation, meaning all deal.II routines operating on individual elements need to compress() after they are done.
3. Instead of calling compress() all over the place, deal.II routines inside and outside the the linear algebra routines should ```Assert(is_compressed())```.

tjhei commented on source/lac/trilinos_sparse_matrix.cc:
```diff
@@ -811,7 +808,7 @@ namespace TrilinosWrappers
       nonlocal_matrix.reset();
 
-    compress();
+    compress(VectorOperation::unknown);
```
I don't think unknown is correct here!

kronbichler replied:
Indeed, VectorOperation::insert is better.

bangerth replied:
OK, I'll gladly trust you on this.

tjhei commented:
I think ```SparseMatrix::clear_row``` needs a compress at the end because the user can call it directly, or does it? Then ```clear_rows``` should call an internal version that doesn't call compress() for every row.

tjhei commented:
I think ```SparseMatrix::compress()``` should not allow the argument ```unknown``` and assert that.

bangerth commented:
Well, funny, how things sometimes develop. I've gone through all of your comments again and found that basically every place you commented on was in fact wrong before. I've updated the documentation in a few places and added a couple of tests as well.

So from my point of view, this is ready for re-review.

tjhei commented:
This looks pretty good to me now, but I would appreciate somebody else to look it over.

@bangerth all the tests in trilinos/ and mpi/ are passing for you?

Also, we should think about adding something similar to ```PETScWrappers::MatrixBase::assert_is_compressed ()``` and sprinkle calls to it in functions like ```clear_row()``` etc..

bangerth commented:
@tjhei : Yes, all tests pass.

As for the `assert_is_compressed()` -- can you open an issue? I think it's a worthwhile thing to have, but I don't think I want to add it for this PR already.

kronbichler commented:
This looks good.
